os-observer/src/actions/fetchGithubIssues/unpaginate.ts

```diff
@@ -0,0 +1,68 @@
+import { graphQLClient } from "./graphQLClient.js";
+
+import { Path, Choose, getPath } from "../../utils/getPath.js";
+
+type PageInfo = {
+  hasNextPage: boolean;
+  endCursor: string;
+};
+
+// TODO: return type should error if not array
+// TODO: pageInfoPath should locate a PageInfo object
+// type HasPageInfo<T extends Record<string | number, any>, K extends Path<T>> = Choose<T, K> extends PageInfo ? K : never;
+
+// typescript can't infer T because it's not passed in (like in getPath), and K can't be inferred precisely without writing
+// out the entire path again. So we curry the function. Sorry for the mess...
+export function unpaginate<T extends Record<string | number, any>>() {
+  return async function <K extends Path<T>, A extends Path<T>>(
+    query: string,
+    dataPath: K,
+    pageInfoPath: A,
+    variables: any = {},
+  ): Promise<Choose<T, K> extends Array<any> ? Choose<T, K> : never> {
+    let cursor = null;
+    const items: any[] = [];
+
+    /* eslint-disable-next-line no-constant-condition */
+    while (true) {
+      const data = await graphQLClient.request<T>(query, {
+        ...variables,
+        cursor: cursor,
+      });
+
+      const pageInfo: PageInfo = getPath(data, pageInfoPath);
+      if (!pageInfo.hasNextPage) {
+        break;
+      }
+
+      const newItems: any[] = getPath(data, dataPath);
+      items.push(...newItems);
+
+      cursor = pageInfo.endCursor;
+
+      const rateLimit: RateLimit = data.rateLimit;
+
+      if (
+        rateLimit.remaining == 0 ||
+        rateLimit.remaining - rateLimit.cost <= 0
+      ) {
+        const timeToReset = Date.parse(rateLimit.resetAt) - Date.now() + 1000;
+        console.log(`sleeping until rate limet reset for ${timeToReset}ms`);
+        await sleep(timeToReset);
```

Comment:
I feel like the beauty of idempotent scripts is you should just be able to mash the button (retry) as often as necessary?

IMO it'd be cleaner to just return with a `rateLimited` and try again, but up to you

Reply:
It really depends where it's being called from. Currently I'm assuming that these function will be called by a script running on github actions, and not burning cycles hammering a timed-out endpoint seems like best practice.

I expect that we will make a couple of small changes when/if we move to initiating the scraping from the client, but I think we can do that when we get there (on demand!).